### PR TITLE
feat: hf trainer examples

### PR DESCRIPTION
**Pull request by garrett361**
## Description

Does several things:
* Adds HF examples under `ds_autotuning_prototype/hf_trainer_api`. Once it's verified that these are good to go, the docs need to be rewritten and these should be merged into `examples/hf_trainer_api`
* Adds sanity checks and forces the user to have a `hyperparameters.deepspeed_config` key
* Adds optional random seed arg and random seed state checkpointing.
* Updated random search logic.

To run any of the HF examples, `cd` into the relevant dir (e.g. `ds_autotuning_prototype/hf_trainer_api/language_modeling`) and run `det e create deepspeed.yaml -i ../det_callback` for a regular single DS trial, or `python3 -m determined.pytorch.deepspeed.dsat deepspeed.yaml -i ../det_callback` for DS AT

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: ds_autotuning_prototype/hf_trainer_api/language-modeling/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+setuptools==59.5.0
+accelerate>=0.12.0
+torch>=1.5.0
+deepspeed==0.9.1
```
Comment (tayritenour) [inline ds_autotuning_prototype/hf_trainer_api/language-modeling/requirements.txt:4]: Is this valid/necessary with our `0.8.3` images?

File: ds_autotuning_prototype/hf_trainer_api/language-modeling/deepspeed.yaml [hunk 2]
```diff
@@ -0,0 +1,44 @@
+name: language-modeling-clm-trainer deepspeed
+debug: false
+environment:
+  environment_variables:
+    - NCCL_DEBUG=INFO
+    # You may need to modify this to match your network configuration.
+    - NCCL_SOCKET_IFNAME=ens,eth,ib
+  image:
+    gpu: determinedai/environments:cuda-11.3-pytorch-1.10-tf-2.8-deepspeed-0.7.0-gpu-9119094
```
Comment (tayritenour) [inline ds_autotuning_prototype/hf_trainer_api/language-modeling/deepspeed.yaml:9]: We could use the `0.8.3` images now

File: ds_autotuning_prototype/hf_trainer_api/image_classification/autotune_config.yaml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  --logging_strategy steps 
+  --logging_steps 10 
+  --evaluation_strategy steps
+  --eval_steps 10
+  --save_total_limit 3 
+  --seed 1337
+  --save_strategy steps
+  --save_steps 20
+  --deepspeed ds_configs/ds_config_stage_1.json
+  --cache_dir ./garrettgoon/cache_dir
```
Comment (tayritenour) [inline ds_autotuning_prototype/hf_trainer_api/image_classification/autotune_config.yaml:39]: This is made on the remote searcher right? If it helps, why would we want to remove it in prod?

File: ds_autotuning_prototype/hf_trainer_api/README.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
+# Vision Transformer with HuggingFace Trainer API and Determined
```
Comment (tayritenour) [inline ds_autotuning_prototype/hf_trainer_api/README.md:1]: Shouldn't this file be in the `image_classification` directory if I'm reading this right?

**Review by tayritenour — APPROVED** [review]
LGTM, minor points that I can just deal with in the clean up

File: ds_autotuning_prototype/hf_trainer_api/language-modeling/requirements.txt (hunk 1 above)
Comment (garrett361) [reply ds_autotuning_prototype/hf_trainer_api/language-modeling/requirements.txt:4]: oh no, I think that was for testing something w/ Cong's example

File: ds_autotuning_prototype/hf_trainer_api/language-modeling/deepspeed.yaml (hunk 2 above)
Comment (garrett361) [reply ds_autotuning_prototype/hf_trainer_api/language-modeling/deepspeed.yaml:9]: yep, will change

Comment (garrett361) [reply ds_autotuning_prototype/hf_trainer_api/language-modeling/deepspeed.yaml]: changed

File: ds_autotuning_prototype/hf_trainer_api/image_classification/autotune_config.yaml (hunk 3 above)
Comment (garrett361) [reply ds_autotuning_prototype/hf_trainer_api/image_classification/autotune_config.yaml]: this whole file shouldn't be there, since we don't need a separate autotuning config anymore. just deleted it.

Not sure what you were asking re: removing something though?

Comment (tayritenour) [reply ds_autotuning_prototype/hf_trainer_api/image_classification/autotune_config.yaml]: the `cache_dir` parameter. The comment below implies we don't want this parameter long term

Comment (garrett361) [reply ds_autotuning_prototype/hf_trainer_api/image_classification/autotune_config.yaml]: oh yeah, here I am specifying some external storage which I mount to my containers on Grenoble. this makes it so you don't have to re-download data sets or model weights for every new trial; they are cached instead.

the path specified here is just very specific to this workflow, so it needs to be removed.

File: ds_autotuning_prototype/hf_trainer_api/README.md (hunk 4 above)
Comment (garrett361) [reply ds_autotuning_prototype/hf_trainer_api/README.md]: did a first pass at updating that readme
